scripts/run-live-build.sh

```diff
@@ -129,36 +129,64 @@ while ! curl --output /dev/null --silent --head --fail \
 done
 set -o errexit
 
+pkg_mirror_secondary=''
 if [[ -n "$DELPHIX_PACKAGE_MIRROR_SECONDARY" ]]; then
-	sed "s|@@URL@@|$DELPHIX_PACKAGE_MIRROR_SECONDARY|" \
-		<config/archives/delphix-secondary-mirror.list.in \
-		>config/archives/delphix-secondary-mirror.list
+    pkg_mirror_secondary="$DELPHIX_PACKAGE_MIRROR_SECONDARY"
+else
+    #
+    # If no secondary package mirror is provided, then pull in the latest
+    # mirror dataset for the build. If no latest dataset is found, then fail.
+    #
+    source_url="http://linux-package-mirror.delphix.com/$UPSTREAM_PRODUCT_BRANCH/latest/"
```

Comment:
Similar the comment I left in the `linux-pkg` review, I wonder if it would be cleaner to keep the Delphix infrastructure-specific details of which mirror URL gets used when building a particular branch to the Delphix-specific jenkins jobs that could then pass in the necessary inputs into this code.

Reply:
I wanted this logic at the appliance-build and linux-pkg level rather than the jenkins level since these workflows can be ran locally without jenkins, so I wanted to support that workflow. I think it would be ideal if the local workflow is as close to prod as possible. However, this logic could potentially be done at the jenkins. Thoughts?

Reply:
One down-side I see of doing it this way is that it makes manual builds without Jenkins harder as we'd have to provide this argument manually. This is especially true for Linux-pkg which is built manually much more often.

Reply:
I can understand that argument. At the same time, we'd be hard coding bits of Delphix-internal infrastructure details within the business logic, which isn't ideal (especially in a public open-source code repo). It's your call, I don't feel that strongly about it.

Reply:
As an operator, I'm biased against hard-coding bits of Delphix-internal infrastructure details within the business logic. But as a developer, I understand that overly onerous environmental setup requirements inhibit code changes. Another option is to have this type of logic in both places: passed-in from `devops-gate` in the general case (which is most friendly to operators) while still making a best effort to infer it outside of `devops-gate` when it is not provided (which is friendly to developers). The more I think about it, the more I'm a fan of this option. What do others think?

Reply:
We already have similar logic in places like [this](https://github.com/delphix/appliance-build/blob/master/scripts/build-ancillary-repository.sh#L45), where we hardcode Delphix-internal infrastructure. I personally don't think that there is sensitive info in there, but adding this will not make things much more coupled to Delphix infrastructure than they are already.

Reply:
I'm fine with @basil's proposal. We already do something like this for the S3 urls in the example I provided above.

Reply:
> At the same time, we'd be hard coding bits of Delphix-internal infrastructure details within the business logic

@sebroy we already do that for certain parts of this repository; e.g. how we download our S3 package artifacts.

Unless we want to couple the build more tightly with Jenkins (e.g. only support building via Jenkins), then this is inevitable due to how we manage some of the things this build depends on. As a design principle, should we only support building via Jenkins? or should we support building outside of Jenkins, and take the approach that Jenkins is just one of the many potential users of the build?

Reply:
> Another option is to have this type of logic in both places: passed-in from devops-gate in the general case (which is most friendly to operators) while still making a best effort to infer it outside of devops-gate when it is not provided (which is friendly to developers).

@basil can you explain how this is different than what we've previously done, and what I think this change is doing? I had thought this was already the approach we should take, in general.

Reply:
Previously we placed direct references to `npm.delphix.com` or `artifactory.delphix.com` in builds. In fact, we have over 5,0000 such references in `dlpx-app-gate` today:

```
[basil:~/work/app/dlpx-app-gate]$ git grep artifactory.delphix.com | wc -l
5642
```

This is friendly to developers but not operators. This PR (as it stands today) continues the same pattern. I am now proposing that we adopt a two-pronged approach where such operational concerns are centralized in `devops-gate` while still making a best effort to infer a sensible value outside of `devops-gate` for convenience during interactive development.

Reply:
In the proposed two-pronged approach (for this particular case), what would the devops-gate code look like? would it effectively be duplicating the logic that's in this PR, of interacting with the "latest" redirect? or something different?

I'm not necessarily against a two-pronged approach in general, but I'm not sure I completely understand the proposed solution without seeing code changes implementing the suggestion.

Reply:
> In the proposed two-pronged approach (for this particular case), what would the devops-gate code look like? would it effectively be duplicating the logic that's in this PR, of interacting with the "latest" redirect?

Yeah, that's exactly what I'm proposing.

Reply:
What is the benefit of duplicating that logic in two repositories?

Reply:
The benefit of keeping the main copy of the logic in `devops-gate` is that it is the most friendly to operators. For example, if the mirror is renamed or the URL scheme changes, the logic in `devops-gate` could be updated and that update would apply to all product branches (including older branches and release tags). If this logic wasn't centralized in `devops-gate`, then there would be increased burden on operators, who would have to maintain the old API contract (e.g. preserving old DNS names and URL schemes) or risk completely breaking builds.

The above is an argument for moving the logic to `devops-gate`, but not for keeping a second copy in product repositories. The argument for doing that _in addition to_ keeping the main copy in `devops-gate` is to facilitate more convenienct interactive builds by making a best effort to infer whatever can be inferred. This best effort should always succeed as long as there are no operational changes. If there are operational changes in the future (such as a change in the DNS name or URL scheme), this best effort might fail (most likely on older branches or tags that wouldn't be updated). But this wouldn't matter so much because the best effort was only intended for convenience, and the primary logic (in `devops-gate`) wouldn't fail.

Reply:
> the logic in devops-gate could be updated and that update would apply to all product branches (including older branches and release tags).

Ah, right; I wasn't thinking about older releases, which can't be updated as we make changes. With that in mind, I definitely agree.

Reply:
Taking the above into account - would it make sense to add this change here, but remove the `exit` if no latest URL is found. Then also add this logic in the devops-gate?